### PR TITLE
Enable FATES harvest using fdyndat data passed from ELM to FATES

### PR DESCRIPTION
**Pull request by aldivi**
ELM code (for harvest control and FATES harvest control) has been updated so that ELM 
input harvest can be passed to FATES to drive FATES harvest code. The ELM namelist 
variables do_harvest and fdyndat can now be used to drive FATES harvest. The FATES 
code has not been modified, as it has already been updated for this purpose. Modifications
to the harvest read and get year functions were required to ensure that the correct harvest
year data are read in. ELM harvest results are unchanged, and FATES harvest matches 
the input based on secondary forest fraction. Branch run control has been updated to enable
branch runs.

[non-BFB] only for FATES and BGC cases

### CODE REVIEW

File: components/clm/src/dyn_subgrid/dynHarvestMod.F90
Comment (rgknox) [inline components/clm/src/dyn_subgrid/dynHarvestMod.F90:120]: was going to check if this code is thread-safe seeing how a global (harvest_rates) is being allocated, but both the initialization and the interp code asserts this is the proc level. 
check

File: components/clm/src/dyn_subgrid/dynTimeInfoMod.F90
Comment (rgknox) [inline components/clm/src/dyn_subgrid/dynTimeInfoMod.F90:147]: this looks fine to me, seeing how the original functionality of this routine is preserved, and should not impact existing calls to it, pinging @bishtgautam

File: components/clm/src/main/clmfates_interfaceMod.F90
Comment (rgknox) [inline components/clm/src/main/clmfates_interfaceMod.F90:304]: thanks for these updates @aldivi

**Review by rgknox — APPROVED** [review]
changes look good

**Comment by bishtgautam** [comment]
@aldivi The first comment of each PR is usually a concise description that is copied/pasted when merging the PR to next/master. Could you edit your first comment to be concise? You can always add additional comments in the PR that describe the changes in more detail. Also, please indicate in your first comment if this PR BFB or non-BFB. An example of the first comment of a PR is https://github.com/E3SM-Project/E3SM/pull/3691. Thanks.

**Comment by aldivi** [comment]
ELM code (for harvest control and FATES harvest control) has been updated so that ELM input harvest can be passed to FATES to drive FATES harvest code. The ELM namelist variables do_harvest and fdyndat can now be used to drive FATES harvest. The FATES code has not been modified, as it has already been updated for this purpose.

This has been tested for all harvest happening in single day in FATES with a global 4x5 data set (landuse.timeseries_4x5_hist_simyr1850-2015_200311.nc) on cori and also a single point dataset (landuse.timeseries_1x1_brazil_hist_simyr1850-2015_c200310.nc) on lobata. Running this new code with ELM harvest gives the same results as running ELM harvest with the corresponding master. Running with FATES harvest applies the correct harvest fraction within FATES, as represented by the amount of land moving from primary to secondary land, and does change biomass/harvest results compared to ELM harvest because FATES is a different vegetation model. Furthermore, FATES has a single-source daily mortality routine that selects the largest daily mortality source, and in some grid cells general mortality is greater than logging mortality, even with a single-day harvest, such that in some cells the prescribed harvest is not applied. FATES fire was not enabled for these tests, as it also competes with the selection of the dominant daily mortality source.

It is important to note that the landuse.timeseries files (for fdyndat) contain model year harvest data in the year+1 data element. This is a legacy format that puts the desired pft and harvest data in the same labelled year because they had been created and read in at the same time. For example, pft data and harvest data are read in at the beginning of year y, and since the pft data represented the state at the beginning of y+1, while the harvest data represented the rate during year y, both were stored in as year y+1 to simplify the storage and reading process. The pft y+1 and harvest y data are also processed at the same time when the land files are created. This storage/labelling of these data was not updated when elm was updated to separate the pft and harvest reading and processing.

So additional modifications to the harvest read and get year functions were required to ensure that the correct harvest year data are read in.

I have also updated some controls to allow branch runs to work.

**Comment by bishtgautam** [comment]
Thanks @aldivi

**Comment by jenniferholm** [comment]
@aldivi 
Thanks for that last description, it was very helpful!

**Comment by bishtgautam** [comment]
@aldivi This PR results in unexpected non-BFB DIFFs in the three non-FATES tests (see https://my.cdash.org/viewTest.php?onlyfailed&buildid=1965321). I will revert this PR from next for now. Please fix the DIFFs by generating baselines and 
comparing your code modifications against those baselines. 

FYI, here are some [notes](https://gist.github.com/bishtgautam/31a73b933e2492698bccfc7b55acf01e) to generate/compare baselines.

**Comment by aldivi** [comment]
@bishtgautam Yes, you will get non-BFB diffs with those tests, which are the transient ones. I couldn't figure out exactly which landuse.timeseries files were used, but if they are like the others they include changing pfts and harvest.
I mischaracterized this as a BFB with ELM-only. This is because I first tested it without the year offset of the input harvest data to ensure that the calculation I moved did not change anything (the sum of the 5 harvest categories). But with the year offset the harvest amounts are different, so there is no BFB behavior overall.

**Comment by bishtgautam** [comment]
@aldivi Few questions for you:
1. Does this PR result in non-BFB for a transient run with SP mode?
2. Does this PR result in non-BFB for a transient run with BGC mode?
3. In this PR, does FATES work with and without BGC? 
4. Does this PR result in non-BFB for transient FATES run with/without BGC?

(cc: @rgknox)

**Comment by aldivi** [comment]
@bishtgautam 

FATES does not run with SP mode, and in this version it does not run with BGC either (FATES and CN cannot both be active). I think that recently FATES has been enabled with the ELM BGC modes, but this PR would have to be merged with the current code to find out how well these play together. It is likely that some conditions will have to be set so that both CN harvest and FATES harvest are not happening at the same time.
So in answer to your questions:
Does this PR result in non-BFB for a transient run with SP mode?
Not applicable
Does this PR result in non-BFB for a transient run with BGC mode?
Not applicable
In this PR, does FATES work with and without BGC?
Only without BGC.
Does this PR result in non-BFB for transient FATES run with/without BGC?
non-BFB for transient FATES run without BGC

**Comment by bishtgautam** [comment]
@aldivi But your PR is causing non-BFB results for (1) transient SP mode; (2)transient BGC mode. see https://my.cdash.org/viewTest.php?onlyfailed&buildid=1965321

**Comment by aldivi** [comment]
Ah, the first two questions were for non-FATES runs.
Yes, the results are non-BFB in BGC mode.
They should also be non-BFB in SP mode.

**Comment by rljacob** [comment]
This can't be merged until after we make the v2.0 tag for watercycle.  attn: @golaz

**Comment by bishtgautam** [comment]
Apart from @rljacob's [comment above](https://github.com/E3SM-Project/E3SM/pull/4024#issuecomment-804383538) regarding the water cycle experiment, I would also like @kvcalvin to approve this PR in regards to non-BFB BGC changes.

**Comment by aldivi** [comment]
@rljacob Does the v2 watercycle use transient runs with harvest? Will there be an additional v2 tag, or is that it?
The issue is that the input harvest data are not currently being applied to the correct year. Fixing this year offset is what causes the non-BFB behavior. So the question is whether we are fine with continuing to apply the input harvest data to the following year, or if we want it to be applied to the correct year. Over the course of a simulation, this may not make a huge difference.

**Comment by rljacob** [comment]
There will be another tag for BGC.  Its up to @golaz if he wants this for WC or not.

**Comment by kvcalvin** [comment]
@aldivi : Just to clarify - all of the non-BFB differences are because wood harvest (and the associated LULC) is shifted by a year (to the correct year). Is that correct?

**Comment by aldivi** [comment]
@kvcalvin Before the time shift of harvest prescription I checked the monthly values over a year for conversion flux, wood C, tot veg c, tot som C, tot ecosys C, and atmospheric co2, and the differences were zero (on the same machine, cori). The only code calculation that was changed in ELM was that the harvest type sum was moved from read-in to the harvest routine, which didn't change the answers.
There is also the possibility of separating the time shift from the fates harvest update.

**Review by kvcalvin — APPROVED** [review]

**Review by thorntonpe — APPROVED** [review]

**Comment by rljacob** [comment]
Notes from telecon:  this is ready to go after v2.0 tag.

**Comment by rljacob** [comment]
@bishtgautam do you think this should be rebased ?

**Comment by bishtgautam** [comment]
Okay

**Comment by rljacob** [comment]
@bishtgautam what is up with this branch?

**Comment by aldivi** [comment]
@rljacob @bishtgautam I have been trying to get the rebased version to run, but continue to encounter bugs, although they do not appear to be related to the new harvest code. Some I have fixed to continue (like a bad pft memory access when the crop model is off), but I have recently just hit a memory failure during land model initialization that needs some effort to track down.
The three automated test failures were with transient land tests that will fail because of the shifted year for input harvest data. Were you running these tests on the non-rebased version? If not, how did you get the rebased version to run?

**Comment by rljacob** [comment]
The tests were run after this branch was merged to next which, after conflicts are resolved, is as good as a rebase.  However that was using next from 27 days ago.

**Comment by aldivi** [comment]
@rljacob ok. This should be the same then because gautam's rebase was 28 days ago. I am going to try with debug mode turned off, although it seems weird that it would run in regular mode if debug mode has a bunch of memory errors associated with the regular code.

**Comment by aldivi** [comment]
@rljacob @bishtgautam Is there still a land-developer test suite? I tried several different land suite names and none of them worked on cori, although last year i did run a land test suite. I want to create some baselines of the pre-rebased commit and compare. I had to make a few code edits to get it running, but I think it is ok. I have to do another run to check the year 2000 harvest because the data was not set correctly.

**Comment by bishtgautam** [comment]
The test suite is `e3sm_land_developer`

**Comment by aldivi** [comment]
@rljacob @bishtgautam 
Ok. I have completed testing on cori. The fates harvest code works correctly, and I did two different baseline comparisons against the oct 7 2021 commit in this branch (just prior to the merge of the fates harvest code). I set the harvest year offset to zero and all of the e3sm_land_developer tests passed. I then set the harvest year offset to 1 in order to read the correct year of input data, and all of the tests passed except for the three transient tests (which is expected):
ERS.ne11_oQU240.I20TRELM.cori-knl_intel
ERS.f19_g16.I20TRGSWCNPRDCTCBC.cori-knl_intel.elm-ctc_f19_g16_I20TRGSWCNPRDCTCBC
ERS.f19_g16.I20TRGSWCNPECACNTBC.cori-knl_intel.elm-eca_f19_g16_I20TRGSWCNPECACNTBC

I did have to make a few mods to get it running, including updating a new namelist check in the configuration, allowing for a non-crop-model pft check if fates is active, and updating the array shape parameter for the harvest data. I have pushed these changes to this rebased branch.
I also ran this with and without debug mode.

**Comment by bishtgautam** [comment]
@aldivi what is the default value for harvest year offset?

**Comment by rljacob** [comment]
@aldivi do you have access to compy?  You should check the exact machine/compiler the ERS test originally failed on.

**Comment by aldivi** [comment]
@rljacob @bishtgautam I have tried to generate the same baselines on compy using both the intel and pgi compilers and they fail to build.
The default offset for the harvest data is 1 year.

**Comment by aldivi** [comment]
@bishtgautam @rljacob It turns out that git does not save/restore submodule tags in the main repo associated with particular branches/commits, and the FATES tag changed in the past month. I realigned the oct 7 2021 master commit (07ff719) with its original FATES tag (36e60bb) and was able to generate baselines on compy. I then made the same comparisons as on cori, with the same results: without the input data year shift all land developer tests pass, and with the shift the three transient tests differ (as expected). I also ran the FATES harvest code on compy and the max difference in secondary forest fraction per grid cell between compy and cori is less than 3x10^-12.

**Comment by aldivi** [comment]
@rljacob @bishtgautam Just checking in on the status of this merge, before master gets too far ahead again. Thanks.

**Comment by rljacob** [comment]
@bishtgautam you can merge this when you're ready.

**Review by jenniferholm — APPROVED** [review]
Approve these changes. Thanks

**Comment by jenniferholm** [comment]
@bishtgautam - is there time to resolve these conflicts and get this PR approved, updated, and merged into master?  This PR has been waiting for a year now, and we need to get these harvesting data code changes integrated.  Other people are waiting on these updates so that they can perform their work.

**Comment by bishtgautam** [comment]
Later today, I will work on merging this to the next while resolving conflicts.

**Comment by bishtgautam** [comment]
@rljacob This PR is non-BFB for transient WC runs. The maint-2.0 hasn't been created. Is it okay to merge this PR?

**Comment by rljacob** [comment]
Since the BGC leads want this, yes.

**Comment by bishtgautam** [comment]
@aldivi @jenniferholm, This is now merged to next.

cc: @rgknox @ckoven

**Comment by aldivi** [comment]
Thanks, Gautam!
Alan

--
Alan Di Vittorio, Ph.D.
Lawrence Berkeley National Laboratory
Climate and Ecosystem Sciences Division
***@***.***

**Comment by rljacob** [comment]
Will this PR cause this test to diff?  SMS.ne30pg2_r05_oECv3.F20TR.compy_intel.eam-wcprod

**Comment by aldivi** [comment]
@rljacob that specific test was not run (only land developer suite was run), but it looks like it is a transient case (F20TR), in which case it will diff

**Comment by aldivi** [comment]
@bishtgautam Any transient run with harvest enabled will diff because the driving harvest data are shifted. This test diffs:
ERS.ne11_oQU240.I20TRELM.cori-knl_intel
The regular I20TRELM compset uses ELM%SP. Does this test use the same compset?

**Comment by rljacob** [comment]
That test, ERS.ne11_oQU240.I20TRELM, is **not** diffing on chrysalis, compy or mappy.

File: components/elm/src/dyn_subgrid/dynSubgridControlMod.F90
Comment (bishtgautam) [inline components/elm/src/dyn_subgrid/dynSubgridControlMod.F90:234]: @aldivi Here is the code that ensures harvest is turned on only when BGC is also turned on. Thus, your modifications to the harvest code should not change SP mode. Do you agree?

**Comment by aldivi** [comment]
@bishtgautam yes, this is correct that harvest is shut off if not using bgc or fates. I thought this was the case, but that particular compset made me pause. Thank you for checking the control logic.

**Pull request merged**
